libraries/AP_Airspeed/AP_Airspeed.cpp

```diff
@@ -399,15 +427,122 @@ void AP_Airspeed::read(void)
+    if (blend_state.enable == BLEND_NONE ||
+        !AP::ahrs().have_inertial_nav()) {
+        process_no_blend();
         return;
     }
-    for (uint8_t i=0; i<AIRSPEED_MAX_SENSORS; i++) {
-        if (healthy(i)) {
-            primary = i;
-            break;
+    
+    // blend the airspeed from several sensors and EKF
+    Vector3f velocity, wind;
+    if (AP::ahrs().get_velocity_NED(velocity) == false) {
+        process_no_blend();
+        return;
+    }
+    wind = AP::ahrs().wind_estimate();
+    velocity -= wind;
+    float airspeed_ekf = velocity.length();
+
+    float airspeed_avg = 0.0;
+    float ekf_sensor_airspeed_diff[AIRSPEED_MAX_SENSORS];
+    uint8_t healthy_sens_num = 0;
+    for (uint8_t i = 0; i < AIRSPEED_MAX_SENSORS; i++) {
+        if (!state[i].healthy) {
+            state[i].blend_healthy = false;
+        }
+        
+        if (state[i].blend_healthy) {
+            airspeed_avg += state[i].airspeed;
+            healthy_sens_num++;
+        }
+    }
+    
+    // calculating difference
+    for (uint8_t i = 0; i < AIRSPEED_MAX_SENSORS; i++) {
+        ekf_sensor_airspeed_diff[i] = fabsf(state[i].airspeed - airspeed_ekf);
+    }
+    
+    if (healthy_sens_num > 0)
+    {
+        airspeed_avg /= healthy_sens_num;
+        
+        const float airspeed_avg_limit = airspeed_ekf * AIRSPEED_BLEND_DIFF_LIMIT;
+
+        // look for healthy sensors that might be failing
+
+        // Idea is to find the sensor which differs much both from other sensors and from ekf value
+        // If there are two sensors then the first check will be true for both, however
+        // the second check will be true only for the sensor which really failed
+        for (uint8_t i = 0; i < AIRSPEED_MAX_SENSORS; i++) {
+            if (state[i].blend_healthy &&
+                    ekf_sensor_airspeed_diff[i] > airspeed_avg_limit && state[i].airspeed > AIRSPEED_BLEND_MIN_SPEED) {
+                gcs().send_text(MAV_SEVERITY_INFO, "Airspeed[%u] sensor blending unhealthy", i);
+                state[i].blend_healthy = false;
+            }
+        }
+
+        // TODO: Check what is really needed and where is used. Maybe something is missed.
+        blend_state.airspeed = 0.0;
+        blend_state.raw_airspeed = 0.0;
+        blend_state.temperature = 0.0;
+        blend_state.last_pressure = 0.0;
+        blend_state.corrected_pressure = 0.0;
+        blend_state.EAS2TAS = 0.0;
+        blend_state.ratio = 0.0;
+        blend_state.offset = 0.0;
+
+        healthy_sens_num = 0;
+
+        for (uint8_t i = 0; i < AIRSPEED_MAX_SENSORS; i++) {
+            if (!state[i].blend_healthy) {
+                continue;
+            }
+            
+            blend_state.airspeed += state[i].airspeed;
+            blend_state.raw_airspeed += state[i].raw_airspeed;
+
+            // TODO: Check for get_temperature result and correction to number of sensors
+            float temperature;
+            get_temperature(i, temperature);
```

Comment:
get_temperature() returns false if it gets updated slower than 10 Hz. So, since we're at the mercy of external UAVCAN sensors updating, it's plausible that it may toggle true/false/true/false on misbehaving external devices. I think you'll need a separate healthy_sens_temperature_num value to track that count. And if it's zero do we want the blend to be zero? Hmm.. this is a weird one..

Reply:
Yes, this one is weird.

Reply:
This has not yet came to UAVCAN...

Reply:
If it's null - then don't touch it